src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,243 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                                shiftUntil = MaxPartialDigits;
+                            }
+                            shiftUntil--;
+                        }
+                        remainingIntDigitCount -= intDigitsSpan.Length;
+
+                        ReadOnlySpan<char> fracDigitsSpan = digitsChunkSpan.Slice(intDigitsSpan.Length);
+                        for (int i = 0; i < fracDigitsSpan.Length; i++)
+                        {
+                            char digitChar = fracDigitsSpan[i];
+                            if (digitChar == '\0')
+                            {
+                                break;
+                            }
+                            if (digitChar != '0')
+                            {
+                                result = default;
+                                return false;
+                            }
+                        }
+                    }
+                    Debug.Assert(bufferInd == -1);
+
+                    unsafe
+                    {
+                        Span<uint> newBuffer = new uint[bufferSize];
+
+                        arrayFromPool = ArrayPool<int>.Shared.Rent(1);
+                        Span<uint> multiplier = MemoryMarshal.Cast<int, uint>(arrayFromPool);
+                        multiplier[0] = TenPowMaxPartial;
+
+                        int blockSize = 1;
+                        while (true)
+                        {
+                            fixed (uint* bufPtr = buffer, newBufPtr = newBuffer, mulPtr = multiplier)
+                            {
+                                uint* curBufPtr = bufPtr;
+                                uint* curNewBufPtr = newBufPtr;
+                                for (int i = 0; i < bufferSize; i += blockSize * 2)
+                                {
+                                    int len = Math.Min(bufferSize - i, blockSize * 2);
+                                    int lowerLen = Math.Min(len, blockSize);
+                                    int upperLen = len - lowerLen;
+                                    if (upperLen != 0)
+                                    {
+                                        BigIntegerCalculator.Multiply(mulPtr, blockSize, curBufPtr + blockSize, upperLen, curNewBufPtr, len);
+                                    }
+
+                                    long carry = 0;
+                                    int j = 0;
+                                    for (; j < lowerLen; j++)
+                                    {
+                                        long digit = (curBufPtr[j] + carry) + curNewBufPtr[j];
+                                        curNewBufPtr[j] = unchecked((uint)digit);
+                                        carry = digit >> 32;
+                                    }
+                                    if (carry != 0)
+                                    {
+                                        while (true)
```

Comment:
Same here, if we could refactor this so that we have `while (curNewBufPtr[j] == 0)`, I think that would greatly help the overall readability here.

Reply:
I came up with some different solutions as follows:
```cs
// 1
do
{
    curNewBufPtr[j]++;
    j++;
}
while (curNewBufPtr[j - 1] == 0);

// 2
while (++curNewBufPtr[j++] == 0) { }

// 3
while (++curNewBufPtr[j] == 0)
{
    j++;
}
```
I prefer solution 3 because condition part of solution 1(`curNewBufPtr[j - 1] == 0`) is not intuitive for me, and solution 2 feels bit complicated. Which one do you think is better?

Reply:
For what it's worth, I think that the current `while (true)` looks better than any of the alternatives.

Reply:
I'm fine with leaving it as is given the complexity around changing it and no clear solution/win.